odl/operator/fspace_ops.py

```diff
@@ -0,0 +1,412 @@
+# Copyright 2014, 2015 Jonas Adler
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Default operators defined on any `FunctionSpace`."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+import numpy as np
+
+# ODL imports
+from odl.operator.operator import Operator
+from odl.set.pspace import ProductSpace
+
+
+__all__ = ('DiscreteGradient', 'DiscreteDivergence')
+
+
+def discrete_part_deriv(f, axis=0, dx=1.0, edge_order=2,
+                        zero_padding=False):
+    """Calculate the partial derivative of `f` along a given `axis` using
+    first-order discrete differences.
+
+    The partial derivative is computed using second order accurate central
+    differences in the interior and either first differences or second order
+    accurate one-sides (forward or backwards) differences at the boundaries.
+
+    Assuming (implicit) zero padding central differences are used on the
+    interior and on endpoints. Otherwise one-sided differences differences
+    are used. Then first-order accuracy can be triggered on endpoints with
+    parameter `edge_order`.
+
+    The returned array has the same shape as the input array `f`.
+
+    Parameters
+    ----------
+    f : `numpy.ndarray`
+         An N-dimensional array
+    axis : `int`, optional
+        The axis along which the partial derivative is evaluated. Default: 0
+    dx : `float`, optional
+        Scalars specifying the sample distances in each dimension `axis`.
+        Default distance: 1.
+    edge_order : {1, 2}, optional
+        Partial derivative is calculated using Nth order accurate
+        differences at the boundaries. Default edge order: 2
+    zero_padding : `bool`, optional
```

Comment:
Should document and throw error if user tries using both zero_padding and edge_order.